File: array/cpp/Candy.cpp
#include <iostream>
#include <vector>

using namespace std;

int candy(int ratings[], int size){
    int left[size+1];
    int right[size+1];
    left[0] = 1;
    right[0] = 1;
    for (int i =1; i<size; i++){
        if (ratings[i] > ratings[i-1]){
            left[i] = left[i-1]+1;
        }else{
            left[i] = 1;
        }
        if (ratings[size-i-1] > ratings[size-i]){
            right[i] = right[i-1]+1;
        }else{
            right[i] = 1;
        }
    }
    int res = 0;
    for (int i = 0; i<size; i++){
        res+=max(left[i],right[size-i-1]);
    }
    return res;
}

int main(){
    int rating[] = {1,0,2};
    int size = sizeof(rating)/sizeof(rating[0]);
    int res = candy(rating,size);
    cout<< "Res: "<< res <<endl;
    return 0;
}

File: revised/problems/com/arvind/revision/greedy/Candy.java
package com.arvind.revision.greedy;

import java.util.Arrays;

public class Candy {

    public int candy(int[] ratings) {
        int[] candies = new int[ratings.length];
        Arrays.fill(candies, 1);

        // left pass
        for (int i = 1; i < candies.length; i++) {
            if (ratings[i] > ratings[i - 1]) {
                candies[i] = candies[i - 1] + 1;
            }
        }

        // right pass
        for (int i = candies.length - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                candies[i] = Math.max(candies[i], candies[i + 1] + 1);
            }
        }

        return Arrays.stream(candies).sum();
    }

    public static void main(String[] args) {
        int[] ratings = {1, 3, 2, 2, 1};
        Candy solution = new Candy();
        System.out.println(solution.candy(ratings));
    }
}
